mypy/semanal.py

```diff
@@ -3936,11 +3936,34 @@ class C:
         #       caught.
         assert self.statement  # we are at class scope
         return (node is None
-                or node.line < self.statement.line
+                or self.is_textually_before_statement(node)
                 or not self.is_defined_in_current_module(node.fullname)
                 or isinstance(node, TypeInfo)
                 or (isinstance(node, PlaceholderNode) and node.becomes_typeinfo))
 
+    def is_textually_before_statement(self, node: SymbolNode) -> bool:
+        """Check if a node is defined textually before the current statement
+
+        Note that decorated functions' line number are the same as
+        the top decorator.
+        """
+        assert self.statement
+        line_diff = self.statement.line - node.line
+
+        # The first branch handles reference an overloaded function variant inside itself,
+        # this is a corner case where mypy technically deviates from runtime name resolution,
+        # but it is fine because we want an overloaded function to be treated as a single unit.
+        if isinstance(node, OverloadedFuncDef) and isinstance(self.statement, FuncDef):
+            return self.is_overloaded_item(node, self.statement) and line_diff > 0
+        elif isinstance(node, Decorator) and not node.is_overload:
+            return line_diff > len(node.original_decorators)
+        else:
+            return line_diff > 0
+
+    def is_overloaded_item(self, overload: OverloadedFuncDef, func: FuncDef) -> bool:
+        """Check whehter the function belongs to the overloaded variants"""
+        return func in overload.items
```

Comment:
This will return `False` in almost all cases, because in most cases items are `Decorator`s, not `FuncDef`. You also need to check `.impl`.